[MHV-44454] Update SM FE to consume Message Thread API changes mhv

vsaleem opened the pull request:
## Summary

- SM FE consumes message thread api updates: message body displays word count of 200 or more and attachments metadata is displayed within the message body
- Accordions are clickable and sends a call to the BE to mark as 'READ'
- When the user clicks expand all, the unread messages within the message thread remain unread unless individually clicked
- All message threads are displayed on the screen; add "5+ more messages') button is removed
- Pressing the 'Print' button displays all expanded message threads in print window and upon closing the print window, the message thread reverts back to the original view before pressing the "Print" button

## Related issue(s)

- _[MHV-44454](https://jira.devops.va.gov/browse/MHV-44454)](https://jira.devops.va.gov/browse/MHV-44454)_ Update front end to consume message thread details API changes

- _[MHV-52790](https://jira.devops.va.gov/browse/MHV-52790)](https://jira.devops.va.gov/browse/MHV-52790)_ Load all messages on the first load of a thread--not batched by 5 at a time

## Testing done

- _Updated unit tests: "Removed unit tests that look for custom print-modal, the custom print-modal was removed and replaced with the browser print window to appear instead. The new unit test confirms that the correct print layout appears on the print.window when the 'print' button is clicked"_
- _E2E updates_
- _Describe the tests completed and the results_
- _Exclusively stating 'Specs and automated tests passing' is NOT acceptable as appropriate testing
- _Optionally, provide a link to your [test plan](https://depo-platform-documentation.scrollhelp.site/developer-docs/create-a-test-plan-in-testrail) and [test execution records](https://depo-platform-documentation.scrollhelp.site/developer-docs/execute-tests-in-testrail)_

## Screenshots

_Note: This field is mandatory for UI changes (non-component work should NOT have screenshots)._

|         | Before | After |
| ------- | ------ | ----- |
| Mobile  |        |       |
| Desktop |        |
![image](https://github.com/department-of-veterans-affairs/vets-website/assets/59583325/c7f2ffd6-479e-4d02-a921-0865892bb3da)
|

## What areas of the site does it impact?

*(Describe what parts of the site are impacted **if** code touched other areas)*

## Acceptance criteria

### Quality Assurance & Testing

- [ ] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ] Linting warnings have been addressed
- [ ] Documentation has been updated ([link to documentation](#) \*if necessary)
- [ ] Screenshot of the developed feature is added
- [ ] [Accessibility testing](https://depo-platform-documentation.scrollhelp.site/developer-docs/wcag-2-1-success-criteria-and-foundational-testing) has been performed

### Error Handling

- [ ] Browser console contains no warnings or errors.
- [ ] Events are being sent to the appropriate logging solution
- [ ] Feature/bug has a monitor built into Datadog or Grafana (if applicable)

### Authentication

- [ ] Did you login to a local build and verify all authenticated routes work as expected with a test user

### :warning: Team Sites (only applies to modifications made to the VA.gov header) :warning:

- [ ] The vets-website header does not contain any web-components
- [ ] I used the [proxy-rewrite steps](https://github.com/department-of-veterans-affairs/vets-website/tree/main/src/applications/proxy-rewrite#local-dev) to test the injected header scenario
- [ ] I reached out in the `#sitewide-public-websites` Slack channel for questions

## Requested Feedback

(OPTIONAL) _What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

Commit:
in progress - working on capturing accordion state to use after printing

Commit:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
[MHV-44454] Updated handle print function to display printThread component only

Commit:
[MHV-44454] Remove printOption; no longer useful... replaced with printThread

Commit:
[MHV-44454] Refactored use of messageHistory and removed isForPrint conditional; no longer needed.

Commit:
[MHV-44454] Added printThread component for print only

Commit:
[MHV-44454] cleanup; removed unused constant; use printThread to handle MessageThreadForPrint component

Asterix22 committed:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Asterix22 committed:
api path updates for e2e tests

Commit:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
[MHV-44454] Updated unit test to render print window

Commit:
cleanup; removed unused props

Commit:
attempt to resolve unit test timeout failure in CL

Commit:
2nd attempt to resolve timeout in CL

Asterix22 committed:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Asterix22 committed:
cypress test fixes

fazilqa committed:
upd: add status code verification

fazilqa committed:
upd: refactor updateDates function in multi-draft tests

Asterix22 committed:
updates for expanded messages e2e tests

Asterix22 committed:
Update secure-messaging-print-messages.cypress.spec.js

Asterix22 committed:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Asterix22 committed:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
[MHV-44454] Cleanup single message thread call to receive full_body text flag

Commit:
[MHV-44454] Unit test for MessageThreadForPrint component and added mock data

- mock data includes full body text and attachment file

Commit:
[MHV-44454] Added helper funcs to handle attachment ids; moved getSize to helper file

Commit:
[MHV-44454] Refactored AssignmentList to use attachment ids for individual

- this was an axe check finding by the automation team; each message thread contains it's own 'RemoveAttachmentModal' and 'RemoveAttachmentButtons' because they live within the AssignmentList component initially. To fix this issue, each id and data-testid needed their own id identifier.

Commit:
[MHV-44454] Updated Unit Test files; added mock file res data for unit test

Commit:
cleanup

Asterix22 committed:
Merge branch '44454-Update-FE-to-consume-Message-Thread-API-changes-mhv' of https://github.com/department-of-veterans-affairs/vets-website into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Asterix22 committed:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

oleksii-morgun-va commented on src/applications/mhv/secure-messaging/components/MessageActionButtons.jsx:
```diff
@@ -25,10 +27,20 @@ const MessageActionButtons = props => {
       type: Actions.Message.SET_THREAD_PRINT_OPTION,
       payload: printOption,
     });
```
I believe you were attempting to cleanup printOption. Do we still need this argument `printOption` and the `dispatch` block?

oleksii-morgun-va commented on src/applications/mhv/secure-messaging/reducers/threads.js:
```diff
@@ -65,3 +65,53 @@ export const threadsReducer = (state = initialState, action) => {
   }
 };
+
+export const messageThreadsReducer = (state = initialState, action) => {
```
Would you please elaborate on the purpose of this reducer export?

vsaleem replied:
This was a mock test for my learning purposes and shouldn't have gone in. Please disregard.😬

oleksii-morgun-va commented on src/applications/mhv/secure-messaging/tests/components/MessageThreadComponents/MessageThreadForPrint.unit.spec.jsx:
```diff
@@ -0,0 +1,61 @@
+    );
+  };
+
+  it('Renders component layout to display extended messages, word count that`s greater than 200, and attachment metadata for print', () => {
```
```suggestion
  it('Renders component layout to display extended messages, character count that`s greater than 200, and attachment metadata for print', () => {
```

alex-morgun-1 reviewed (changes requested):
Hey Vic, this is enormous work. I left my review comments. Outside of that, we need to look at the redundancy of these api calls, some are duplicates and others are not handled properly
![image](https://github.com/department-of-veterans-affairs/vets-website/assets/87077843/5b5cc451-0ea9-419e-b980-5e08b85b6d91)

Commit:
cleanup

Commit:
Merge branch '44454-Update-FE-to-consume-Message-Thread-API-changes-mhv' of https://github.com/department-of-veterans-affairs/vets-website into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
[MHV-44454] Cleanup; Removed printOption prop to fully rely on printThread to set print; cleanup unit tests

Commit:
[MHV-44454] Removed unused reducer

Commit:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
[MHV-44454] Updated retrieveMessageThread to make one call to api and updated errors dispatch

Commit:
[MHV-44454] Moved RemoveAttachmentModal outside of AttachmentsList component

Commit:
[MHV-44454] Updated accordion to expand on 'unread' messages only; removed preloaded; removed expandListener callback

Commit:
cleanup

Asterix22 committed:
more e2e test fixes

oleksii-morgun-va commented on src/applications/mhv/secure-messaging/components/AttachmentsList.jsx:
```diff
@@ -113,45 +25,6 @@ const AttachmentsList = props => {
-            </button>
-          </VaAlert>
-        )}
-
```
I'm trying to understand the reasoning behind this change. I'm wondering if it was intentional to distribute one piece of code across four instances in two separate components. I'm open to different approaches if it is absolutely necessary. let me know if I am missing something here

alex-morgun-1 committed:
[#MHV-44454] adding print-only class on print thread

christinec-fftc commented on src/applications/disability-benefits/all-claims/subtask/pages/appeals.jsx:
```diff
@@ -77,7 +77,6 @@ const AppealsPage = ({ data = {}, error, setPageData }) => {
             value={value}
             label={label}
             checked={value === data[dataKey]}
-            uswds={false}
```
Looks like this pr is reverting changes from #27683 and #27724. Is that intended?

vsaleem replied:
Hi @christinec-fftc this change is not intended. I'm working on cleaning up this branch and adding these changes back shortly

pmclaren19 commented on src/applications/appeals/10182/containers/ConfirmationPage.jsx:
```diff
@@ -44,7 +44,7 @@ export const ConfirmationPage = () => {
         />
         <h2>Request for Board Appeal</h2>
       </div>
-      <va-alert status="success" ref={alertRef} uswds>
```
Why are you removing the `uswds` prop? if you no longer want this prop enabled you should do `uswds="false" `since this prop will be enabled after Feb 16th.

vsaleem replied:
Hi @pmclaren19 this change is not intended. I'm working on cleaning up this branch and adding these changes back shortly

pmclaren19 reviewed (dismissed):
Looks like your removing `uswds` prop from a lot of va components. I dont see anything in your ticket about this ...so Im assuming this is a mistake and these changes need to be added back? If you no longer want this prop enabled you should do `uswds="false" `since this prop will be enabled after Feb 16th.

Commit:
[MHV-44454] Moved RemoveAttachmentModal into AttachmentsList component

- Added 'compose' and 'reply' props, RemoveAttachmentModal only appears in the dom if either props are true

Commit:
[MHV-44454] Removed printThread prop; Updated ids for 'print-only' and screen

- MessageThreadForPrint component is only displayed when using the 'print-only' css but is still displayed in the DOM.

- Updated ids to identify forPrint component and not-for-print components.
- Updated each messageThread item component to have its own id, (message-body, headers, etc.)

Commit:
[MHV-44454] Unit tests

Asterix22 committed:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

vsaleem commented:
> Looks like your removing `uswds` prop from a lot of va components. I dont see anything in your ticket about this ...so Im assuming this is a mistake and these changes need to be added back? If you no longer want this prop enabled you should do `uswds="false" `since this prop will be enabled after Feb 16th.

The above was done in error and everything should be back to normal. Thank you for taking a look at this!  Please take another look whenever you have a moment @pmclaren19

Asterix22 committed:
Update secure-messaging-message-thread-details.cypress.spec.js

Asterix22 committed:
Update secure-messaging-message-thread-details.cypress.spec.js

Asterix22 committed:
Update secure-messaging-message-thread-details.cypress.spec.js

Asterix22 committed:
Update secure-messaging-message-thread-details.cypress.spec.js

Commit:
[MHV-44454] Updated condition to make one call to mark message as read

Commit:
Merge branch '44454-Update-FE-to-consume-Message-Thread-API-changes-mhv' of https://github.com/department-of-veterans-affairs/vets-website into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
[MHV-44454] Cleanup attachmentName IDs; using forPrint prop to identify ids

Commit:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Asterix22 committed:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
[MHV-44454] Update id name and unit test

Asterix22 committed:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Asterix22 committed:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
attempt to resolve flaky test

Commit:
Merge branch '44454-Update-FE-to-consume-Message-Thread-API-changes-mhv' of https://github.com/department-of-veterans-affairs/vets-website into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Commit:
Merge branch 'main' into 44454-Update-FE-to-consume-Message-Thread-API-changes-mhv

Asterix22 committed:
fixes to e2e

Commit:
cleanup unit tests

oleksii-morgun-va commented on src/applications/mhv/secure-messaging/components/AttachmentsList.jsx:
```diff
@@ -232,19 +246,21 @@ const AttachmentsList = props => {
-        }}
-        data-testid="remove-attachment-modal"
-      />
+      {(compose || reply) && (
```
not substantial, but appears there is no difference in the use case of these 2 props, I would personally go with with 1 prop like `isEdit` or something. Unless we are planning to have separate conditions for each of these props.

vsaleem replied:
These props are used independently on separate pages, compose and reply screens. One or the other would have to be true for the modal to be used.

oleksii-morgun-va replied:
right, but they end up both doing the same thing within the component. You can technically still send like `isEdit` prop from both pages without sacrificing any functionality. I guess it is just a matter of preference and keeping code less confusing

oleksii-morgun-va commented on src/applications/mhv/secure-messaging/tests/components/MessageThreadComponents/MessageThreadForPrint.unit.spec.jsx:
```diff
@@ -0,0 +1,65 @@
+      ).textContent,
+    ).to.equal(`${attachmentName}(${getSize(attachmentSize)})Has attachment`);
+  });
+});
```
since the component accepts 2 props, do we need to add tests that handle scenarios when props are invalid?

alex-morgun-1 reviewed (approved):
Looks great. approved with 2 suggestions
